### PR TITLE
hints: better clicking position

### PR DESCRIPTION
**Pull request by lahwaacz**
I've used the `getClientRects()` JavaScript method to get a more reliable position for the emulated click in hint mode. The previous method (center of the `<a>` element) is used as a fallback, just in case that all the obtained bounding boxes have zero dimensions (or when none were returned?).

I'm keeping the changes as minimal as possible to increase the chances of this request being accepted :stuck_out_tongue: and I've also added tests for the cases I expect this will fix (issues #292, #1298, #1304). Let me know if I should change/add/remove something.

[Review on Reviewable](https://reviewable.io/reviews/the-compiler/qutebrowser/1311)

### CODE REVIEW

**Comment by The-Compiler** [comment]
Thank you! Since it comes with tests, I'm indeed somewhat inclined to merge it sooner rather than later :wink:

I started on a different way of doing hint tests on Friday (with a simple folder with HTMLs, and a special comment in the HTML file saying what hints should be found), but I'm not finished yet. I guess we can still convert your test cases to that format later if it makes sense.

Some more things I'd like to see tested:

- Running all tests with Javascript turned off (to get the old `rect_on_view` behaviour). Some tests probably need pytest mark to be marked as expected failure in that case, I can do that part if you don't want to figure it out.
- Hints in an iframe
- A scrolled page
- A scrolled iframe
- Both a scrolled page and a scrolled iframe

Also see `TestRectOnView` in `tests/unit/browser/test_webelem.py`. (Now that I think of it, maybe this should actually be implemented there?)

File: qutebrowser/browser/hints.py [hunk 1]
```diff
@@ -438,14 +464,24 @@ def _click(self, elem, context):
...
         # https://github.com/The-Compiler/qutebrowser/issues/70
-        pos = elem.rect_on_view().center()
+        rect = self._get_first_rectangle(elem)
+        pos = rect.center()
+
+        # fix coordinates according to zoom level
+        zoom = elem.webFrame().zoomFactor()
+        if not config.get('ui', 'zoom-text-only'):
+            pos.setX(pos.x() * zoom)
+            pos.setY(pos.y() * zoom)
```
Comment (The-Compiler) [inline qutebrowser/browser/hints.py]: I think this accounts for the zoom twice if `_get_first_rectangle` falls back on `rect_on_view`?

**Comment by lahwaacz** [comment]
Is it possible to somehow describe how many times and with which settings a test should be run or should I simply duplicate each scenario?

As for disabling JavaScript, that's not what the code currently does. Even when I `:set content allow-javascript false`, the `elem.evaluateJavaScript` method still works. I could check the config manually, but I'm not even sure if `allow-javascript` should apply to `evaluateJavaScript`?

**Comment by The-Compiler** [comment]
Huh, interesting. Officially, that shouldn't work (and `QWebFreame.evaluateJavaScript` indeed doesn't!). When I tried some months back, I noticed the same as you did, but when I tried again recently it didn't work anymore.

I can't think of a clean way for the JS/non-JS tests right now, but I'll take care of that with the test layout I mentioned earlier, I think I'll have time to implement that on Monday (or maybe earlier).

**Comment by lahwaacz** [comment]
Perhaps it would be good to have a separate setting for the "internal" JavaScript code (possibly inheriting from `allow-javascript` by default). It is generally much lighter and more "trusted" than the regular content, so it might make sense to disable `allow-javascript` while leaving the browser's functionality unimpaired.

**Comment by The-Compiler** [comment]
Ideally internal javascript would always run, no point in being able to turn it off anyways. However, Qt (or at least QtWebKit) doesn't allow this to be controlled separately. In theory, as soon as you turn off JS for a website, running JS via the C++/Python API won't work anymore either (but as we've seen, *sometimes* it does for some reason).

**Comment by lahwaacz** [comment]
Huh, after the fix in 1cff16b the `py34-cov` and `unittests-nodisp` testsuites failed on Travis (caused by the zoom tests), but `py35` run fine...

**Comment by The-Compiler** [comment]
Note `py35` runs on OS X (with a more recent Qt), so that might also be something different between Linux/OS X, or the Qt versions. No idea what that could be though...

**Comment by The-Compiler** [comment]
I can reproduce the issue on Ubuntu Trusty (Qt 5.2.1 with QtWebKit 5.1) but not on Debian Jessie (Qt 5.3.2).


Good log:

```
commands   command:run:504 Calling qutebrowser.browser.hints.HintManager.start(<qutebrowser.browser.hints.HintManager object at 0x7f2d9051e4c8>, False, <Group.links: 2>, <Target.normal: 1>, win_id=0)
hints      hints:_init_elements:717 hints: a, s, d, f, g, h
[...]
commands   command:run:504 Calling qutebrowser.browser.hints.HintManager.follow_hint(<qutebrowser.browser.hints.HintManager object at 0x7f2d9051e4c8>, 'h')
modes      modeman:leave:285 Leaving mode KeyMode.hint (reason: followed)
hints      hints:_cleanup:176 Disconnecting frame <PyQt5.QtWebKitWidgets.QWebFrame object at 0x7f2d9051e558>
hints      hints:_cleanup:190 Disconnected.
statusbar  text:maybe_reset_text:74 Resetting: 'Follow hint...'
statusbar  text:set_text:59 Setting normal text to ''.
hints      hints:_get_first_rectangle:440 Client rectangles of element '<a href="/data/hello.txt">id</a>': {'item': {}, 'length': 1.0, '0': {'width': 12.0, 'top': 539.0, 'bottom': 553.0, 'right': 368.0, 'height': 14.0, 'left': 356.0}}
hints      hints:_click:484 Clicking on '<a href="/data/hello.txt">id</a>' at position PyQt5.QtCore.QPoint(361, 545)
webview    webpage:on_start_hinting:450 Setting force target to ClickTarget.normal
modes      webview:on_mouse_event:404 Hint triggered, focusing <qutebrowser.browser.webview.WebView tab_id=0 url='http://localhost:42944/data/hints/link.html'>
modes      webview:on_mouse_event:404 Hint triggered, focusing <qutebrowser.browser.webview.WebView tab_id=0 url='http://localhost:42944/data/hints/link.html'>
modes      webview:on_mouse_event:404 Hint triggered, focusing <qutebrowser.browser.webview.WebView tab_id=0 url='http://localhost:42944/data/hints/link.html'>
misc       webelem:is_editable:261 Checking if element is editable: <qutebrowser.browser.webelem.WebElementWrapper html='<a href="/data/hello.txt">id</a>'>
mouse      webview:_mousepress_insertmode:282 Clicked non-editable element!
modes      modeman:maybe_leave:113 Not in mode KeyMode.insert! (leave reason: click)
mouse      webview:_mousepress_opentarget:325 Normal click, setting normal target
[...]
webview    webview:_set_load_status:182 load status for <qutebrowser.browser.webview.WebView tab_id=0 url='http://localhost:42944/data/hello.txt'>: LoadStatus.success
```

Bad log:

```
commands   command:run:504 Calling qutebrowser.browser.hints.HintManager.start(<qutebrowser.browser.hints.HintManager object at 0x7f81d4a7e708>, False, <Group.links: 2>, <Target.normal: 1>, win_id=0)
hints      hints:_init_elements:717 hints: a, s, d, f, g, h
[...]
commands   command:run:504 Calling qutebrowser.browser.hints.HintManager.follow_hint(<qutebrowser.browser.hints.HintManager object at 0x7f81d4a7e708>, 'h')
modes      modeman:leave:285 Leaving mode KeyMode.hint (reason: followed)
hints      hints:_cleanup:176 Disconnecting frame <PyQt5.QtWebKitWidgets.QWebFrame object at 0x7f81d4a7e798>
hints      hints:_cleanup:190 Disconnected.
statusbar  text:maybe_reset_text:74 Resetting: 'Follow hint...'
statusbar  text:set_text:59 Setting normal text to ''.
hints      hints:_get_first_rectangle:440 Client rectangles of element '<a href="/data/hello.txt">id</a>': {'0': {'bottom': 584.0, 'width': 13.0, 'top': 569.0, 'left': 345.0, 'right': 358.0, 'height': 15.0}, 'item': {}, 'length': 1.0}
hints      hints:_click:484 Clicking on '<a href="/data/hello.txt">id</a>' at position PyQt5.QtCore.QPoint(351, 576)
webview    webpage:on_start_hinting:450 Setting force target to ClickTarget.normal
modes      webview:on_mouse_event:404 Hint triggered, focusing <qutebrowser.browser.webview.WebView tab_id=0 url='http://localhost:44760/data/hints/link.html'>
modes      webview:on_mouse_event:404 Hint triggered, focusing <qutebrowser.browser.webview.WebView tab_id=0 url='http://localhost:44760/data/hints/link.html'>
modes      webview:on_mouse_event:404 Hint triggered, focusing <qutebrowser.browser.webview.WebView tab_id=0 url='http://localhost:44760/data/hints/link.html'>
mouse      webview:_mousepress_insertmode:252 Clicked at PyQt5.QtCore.QPoint(351, 576) but frame is None!
mouse      webview:_mousepress_opentarget:325 Normal click, setting normal target
webview    webpage:on_stop_hinting:456 Finishing hinting.
```

**Comment by The-Compiler** [comment]
You got the right hunch in https://github.com/The-Compiler/qutebrowser/pull/1313#issuecomment-187127722 - the element is *just* out of the viewport in the Ubuntu VM. For some reason, it's still counted as "visible" (and thus gets a hint) but it's not clickable.

In the Debian VM (and I'm guessing on OS X on Travis as well), the text is rendered slightly smaller, and thus clicking works.

**Comment by The-Compiler** [comment]
I'd suggest splitting the tests into individual, stand-alone HTML pages - that also makes it much easier to repeat tests manually as you know what to do.

I've started implementing the tests based on example files in 36b0e304fc46cd6821be1804bc8f9fae280b9cef - see the commit message on how they work. You can simply drop HTML files into `tests/integration/data/hints/html/` and add a `<!-- target: hello.txt -->` comment. Then `tests/integration/test_hints_html.py` will pick it up, check there's exactly one hint, and that it points to that page when clicked.

This should help simplifying the easier cases - we can still use BDD-style tests for things which require scrolling/zooming for now.

**Comment by lahwaacz** [comment]
That's the problem with zooming tests. On one hand we need to have a link near the bottom-right corner so that changing the zoom level moves the link visually to a completely different place, but on the other hand we run into the out-of-view problems on _some_ systems.

But splitting the tests is definitely a good idea, I'll look into that.

**Comment by The-Compiler** [comment]
Another possibility would probably be to do a `:scroll bottom` before hinting, no?

**Comment by lahwaacz** [comment]
The only explanation for why the tests passed before 1cff16b3b4b709f417766836b4fc966fe6c63952 and failed after are roudning errors. They influence also other things, e.g. d630f966e7a3e680018dc362f67ebc53e8bc8c90 is a fix for a problem happening at 66% and 33% zoom levels, but not at 25%, 50% nor 75%.

**Comment by The-Compiler** [comment]
Good catch! Are there already testcases for those weird corner-cases where possible? If not, please add them, as I'm sure this is going to bite us in the future again otherwise :wink:

**Comment by lahwaacz** [comment]
I've added parametrization of the zoom level in 891a31a0cacc26fd867199769636f5360dcc5411.

**Comment by The-Compiler** [comment]
Great, thanks! So what's still missing here from my POV:

- Tests for scrolling/iframes (like mentioned above)
- Moving this to `webelem.rect_on_view` (if it does the same but better, which I think it does?) and adjusting related unit tests so it's also used for e.g. calculating the display position of the hints, which I think it isn't currently?

Please let me know if you need help! I'll also look into writing a standalone test to investigate the weird behaviour around `evaluateJavaScript` with JS turned off.

**Comment by lahwaacz** [comment]
Moving the `getClientRects` method directly to `webelem.rect_on_view` is an interesting idea, but it would mean many problems - especially in the areas of code I'm not acquainted with.

Problems with positioning hints according to `getClientRects`:
- First problem is performance, on pages with many links it is noticeably slower than the original method.
- Second problem are zoom levels (again). The best I've been able to come up with is this:
```patch
diff --git a/qutebrowser/browser/hints.py b/qutebrowser/browser/hints.py
index ef2f37f..3390d7a 100644
--- a/qutebrowser/browser/hints.py
+++ b/qutebrowser/browser/hints.py
@@ -349,7 +349,7 @@ class HintManager(QObject):
             ('display', 'inline !important'),
             ('z-index', '{} !important'.format(int(2 ** 32 / 2 - 1))),
             ('pointer-events', 'none !important'),
-            ('position', 'absolute !important'),
+            ('position', 'fixed !important'),
             ('color', config.get('colors', 'hints.fg') + ' !important'),
             ('background', config.get('colors', 'hints.bg') + ' !important'),
             ('font', config.get('fonts', 'hints') + ' !important'),
@@ -375,15 +375,11 @@ class HintManager(QObject):
             elem: The QWebElement to set the style attributes for.
             label: The label QWebElement.
         """
-        rect = elem.geometry()
+        rect = self._get_first_rectangle(elem)
         left = rect.x()
         top = rect.y()
-        zoom = elem.webFrame().zoomFactor()
-        if not config.get('ui', 'zoom-text-only'):
-            left /= zoom
-            top /= zoom
-        log.hints.vdebug("Drawing label '{!r}' at {}/{} for element '{!r}', "
-                         "zoom level {}".format(label, left, top, elem, zoom))
+        log.hints.vdebug("Drawing label '{!r}' at {}/{} for element '{!r}'"
+                         .format(label, left, top, elem))
         label.setStyleProperty('left', '{}px !important'.format(left))
         label.setStyleProperty('top', '{}px !important'.format(top))

@@ -436,7 +432,7 @@ class HintManager(QObject):
             elem: The QWebElement of interest.
         """
         rects = elem.evaluateJavaScript("this.getClientRects()")
-        log.hints.debug("Client rectangles of element '{}': {}"
+        log.hints.vdebug("Client rectangles of element '{}': {}"
                 .format(elem.debug_text(), rects))
         for i in range(int(rects.get("length", 0))):
             rect = rects[str(i)]
```
I had to change `position:absolute` to `position:fixed` to make it work with scrolling and because zoom is treated differently, it is incompatible with the fallback solution using `rect_on_view`. Maybe I just don't fully understand the coordinate systems involved...

**Comment by lahwaacz** [comment]
Hmm, the `getClientRects` method surprisingly takes care of scroll offset, but not of parent frames (fixed in b1207650b08639e0d852fa8aae4290417eb77a0b).

I've started working on the iframe tests, but can't figure out how to check if the target page has been loaded correctly. I've tried `Then data/hello.txt should be loaded` as before, but it seems that the expected message is not logged when the page is loaded inside an iframe:

    webview:_set_load_status:182 load status for <qutebrowser.browser.webview.WebView tab_id=0 url='http://localhost:53234/data/hello.txt'>: LoadStatus.success

For now I've xfailed the tests, but please take a look at this.

**Comment by The-Compiler** [comment]
Will do, but I can't promise I'll have time this week, kind of busy currently.

**Comment by The-Compiler** [comment]
Hmm, the best solution I can see right now is checking for the `acceptNavigationRequest` message:

```diff
diff --git a/tests/integration/features/hints.feature b/tests/integration/features/hints.feature
index 13346aa..8b58ad3 100644
--- a/tests/integration/features/hints.feature
+++ b/tests/integration/features/hints.feature
@@ -12,16 +12,14 @@ Feature: Using hints
 
     ### iframes
 
-    @xfail
     Scenario: Using :follow-hint inside an iframe
         When I open data/hints/iframe.html
         And I run :hint all normal
         And I run :follow-hint a
         And I run :hint links normal
         And I run :follow-hint a
-        Then data/hello.txt should be loaded
+        Then "acceptNavigationRequest, url http://localhost:*/data/hello.txt, type NavigationTypeLinkClicked, *" should be logged
 
-    @xfail
     Scenario: Using :follow-hint inside a scrolled iframe
         When I open data/hints/iframe_scroll.html
         And I run :hint all normal
@@ -29,4 +27,4 @@ Feature: Using hints
         And I run :scroll bottom
         And I run :hint links normal
         And I run :follow-hint a
-        Then data/hello2.txt should be loaded
+        Then "acceptNavigationRequest, url http://localhost:*/data/hello2.txt, type NavigationTypeLinkClicked, *" should be logged
```

**Comment by The-Compiler** [comment]
Hey @lahwaacz, any update on the `iframe` tests?

**Comment by lahwaacz** [comment]
Sorry for the delay, I got a bit busy in the meantime and somewhat forgot about this...

The checks above look a bit crazy, but seem to work fine, so I've used them in the tests until there is better solution. I've also added another test for opening a link in a specific iframe, not sure if we should test the other [targets](http://www.w3schools.com/tags/att_a_target.asp) as well. So the existing tests seem to be working fine, I don't know what else should be tested - perhaps something crazy like nested iframes or common Javascript solutions?

**Comment by lahwaacz** [comment]
I've merged the master branch into the PR today, resolved a conflict in `tests/integration/features/hints.feature` and made a small adjustment to fix the tests that were added in the last month. Since all tests added here are passing and the PR seems to be open due to some still missing tests, I think it's time to finally merge this to let users benefit from the fixes. The missing tests can be documented in a (new) issue and added later. I will be happy to work on this further when I have more free time. I haven't had much of it for a few months and the exam period is slowly coming, sorry...

**Comment by The-Compiler** [comment]
Thanks for the update! I agree with you, sorry for being a bit too pedantic maybe. It's definitely time to get this out.

I'm trying to merge various PRs lately to get the count of open PRs down a bit, and I'll take care of this one as soon as possible too.

I also plan to take another look at #1194 soon and either update and merge it, or pick individual fixes out of it and add them while adding tests for them too.

Thanks again for the patience, and best of luck with your exams!

**Comment by lahwaacz** [comment]
As for #1194, the changes are working perfectly for me but I never got to writing the tests for it. Can do a rebase if you wish.

And there are actually two more commits that are not present in any PR yet: [the first](https://github.com/lahwaacz/qutebrowser/commit/4bf88900e14abce53b4ce6c3c0901c350b62e6ad) moves the click position from the center of the element near the top-left corner, which should help if the element is overlayed with another one (there is an issue for this that I can't remember at the moment). [The second](https://github.com/lahwaacz/qutebrowser/commit/62728316b5156c369d6faca6964b6b9113d75e4f) uses the `getClientRects()` method also for positioning the hint label, which is visually more pleasing but also considerably slower on pages with many links, so I'm not sure if it should be used by default. I would create a PR for this eventually, but it would be nice if you considered it along with the other PRs.

**Comment by The-Compiler** [comment]
I'd be glad if you could open separate PRs as-is for those, or add them to #1194 . While I can't promise I'll get to them in the next few days, that'll make sure things don't get lost. I can take care of adding tests for them.

I'd be glad if you could rebase #1194, but if you'd rather have me do it, that's okay as well :wink:

**Comment by lahwaacz** [comment]
I've resolved the conflict in #1194 by merging the master branch, since that's only one conflict (rebase would mean resolving all previous conflicts again). The two commits mentioned in my previous comment are in a separate PR #1433 so that you can merge the others independently.

**Comment by The-Compiler** [comment]
Thank you! You might want to set up [git rerere](https://git-scm.com/blog/2010/03/08/rerere.html) so git remembers previous conflict resolutions.

**Comment by The-Compiler** [comment]
Sorry for the delay, and thanks for the contribution - merged!

**Pull request merged**